storage/innobase/fil/fil0crypt.cc

```diff
@@ -1305,16 +1351,48 @@ static bool fil_crypt_space_needs_rotation(rotate_thread_t *state,
     key_state->key_version = crypt_data->encrypting_with_key_version;
   }
 
+  mutex_enter(&crypt_data->mutex);
+
   /* If used key_id is not found from encryption plugin we can't
   continue to rotate the tablespace */
+
   if (!crypt_data->is_key_found()) {
-    return false;
+    // We can end up here in case we try to encrypt tablespace but the key used
+    // by this tablespace is no longer in keyring. This can happen when keyring
+    // was changed or crypt_data is in version 1 and key's uuid is empty.
+    if (crypt_data->rotate_state.active_threads == 0 &&
+        crypt_data->encryption == FIL_ENCRYPTION_DEFAULT) {
+      ut_ad(
```

Comment:
1. Because tablespace could be first encrypted then decrypted - i.e. we have unencrypted tablespace with crypt_data in page0 and key_id assigned in page0. Assume the scenario that all tablespaces in the server were encrypted, then decrypted and user removed the keyring. After some time the user decides to encrypt the tablespace again - so we need to create the necessary keys again. Throwing an error here would not be good option - why user would need to retain old keyring if he decrypted all the tablespaces?
2. crypt_data can be v1 without encryption='n'. Remember that encryption threads works in earlier versions of 8.0.

Reply:
Agree with the intention but not with the approach or design. Once a tablespace is decrypted, why do we care about crypt_data? We do we even load crypt_data? 

Can't we just crypt_data when it was decrypted? Looks like we retaining some 'state' even after it is decrypted.

On 2)

Didn't we agree that we will not handle v1 format at all (except that somehow handle encryption='n' tables somehow)?

Reply:
Overall the design came from MariaDB. In details:
1) crypt_data can also contains encryption key id - we retain this id.
2) This is how we handle encryption='n' - by having crypt_data with information that this table should be excluded by encryption threads.

Reply:
ok, this is similar to my above concern to get rid of handling of v1 keys (the goal).
As per IM discussion, we only need to handle v2 keys. Then it is fine.